test/e2e/network/endpointslice.go

```diff
@@ -521,7 +521,11 @@ func hasMatchingEndpointSlices(cs clientset.Interface, ns, svcName string, numEn
 	for _, endpointSlice := range esList.Items {
 		actualNumEndpoints += len(endpointSlice.Endpoints)
 	}
-	if actualNumEndpoints != numEndpoints {
+	// In some cases the EndpointSlice controller will create more
+	// EndpointSlices than necessary resulting in some duplication. This is
+	// valid and tests should only fail here if less EndpointSlices than
+	// expected are added.
+	if actualNumEndpoints < numEndpoints {
```

Comment:
Sadly I think we need something a bit more advanced here. As an example, we do need to ensure that we're not just creating endpoints for every pod in the cluster. Duplicate endpoints are fine, but endpoints referring to Pods that are not in this Service are not, I don't think we have sufficient coverage for that yet. I think we previously got around this with a check to ensure that all Pods we expected were referenced with Endpoints. That combined with this check to ensure there were no excess endpoints was sufficient, now it's not.

Reply:
I think wojtek added some helper recently along this lines, will check tomorrow,

Reply:
is not all deduplicationg and matchin addresses done in ?
https://github.com/kubernetes/kubernetes/blob/beef6ae51245515ed9c68090b9288d986a882c7b/test/e2e/network/endpointslice.go#L314-L317

IIUIC that functions calls this `hasMatchingEndpointSlices()` just to return the slices, and then it seems to do all dedup that you mention

Reply:
Sorry for the delay on this! I think there's one more place that needs to change:

https://github.com/kubernetes/kubernetes/blob/master/test/e2e/network/endpointslice.go#L548-L555

Currently `numIPs` is based on `len(pods)`. If we changed that logic to instead count the number of unique IPs with something like a set I think that would fix the flakiness here.

Reply:
but these are Endpoints, not Slices, len(pods) has to match the number of addresses, the Endpoints doesn't have duplicates, right?

Reply:
🤦 you're completely right. Maybe what we'd run into there would be one older EndpointSlice with fewer endpoints than the newer one. Even still, this should fix the majority of flaky tests.